Add Blogs I Follow Widget

themoonisdown opened the pull request:
This PR (currently still in progress) ports the Blogs I Follow widget into the Jetpack plugin.
https://en.support.wordpress.com/widgets/blogs-i-follow-widget/
cc @mdawaffe 

-------------------
- [ ] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [ ] Did you make changes, or create a **new .js file**? If [Gulp](http://gulpjs.com/) is installed on your testing environment, run `gulp js:hint` before to commit your changes. It will allow you to [detect errors in Javascript files](http://jshint.com/about/).
- [ ] Did you create a **new action or filter**? [Add inline documentation](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#4-hooks-actions-and-filters) to help others understand how to use the action or the filter.
- [ ] Create **[unit tests](https://github.com/Automattic/jetpack/tree/master/tests)** if you can. If you're not familiar with Unit Testing, you can check [this tutorial](https://pippinsplugins.com/series/unit-tests-wordpress-plugins/).

themoonisdown commented on modules/widgets/blogs-i-follow/interface-blogs-i-follow-datastore.php:
```diff
@@ -0,0 +1,68 @@
+			'user_id' => JETPACK_MASTER_USER,
+			'method' => 'GET',
+		);
+		$response = Jetpack_Client::remote_request( $request_args );
```
I have yet to determine why this doesn't work on a Jetpack-connected site. I consistently get the error: "An active access token must be used to query information about the current user." So for some reason even though Jetpack is authenticated to WP.com it doesn't have the token.

Querying the API manually works (using my own token) so temporarily I copied that data into the file (below) so I could continue making forward progress.

mdawaffe commented:
I know you can't see the WordPress.com code, but can you tell how the other widgets handle sharing code between Jetpack and WordPress.com?

What are your plans for blavatars (site icons) and blog descriptions?

themoonisdown commented:
The only things I see are techniques like `if ( defined( 'IS_WPCOM' ) && IS_WPCOM ) {` or `function_exists`. Is there another way that I'm missing? I don't see a lot of special handling of this. I'm happy to scrap the interface I used initially. It was worth a shot, but is over-engineered and as it's so drastically different from the other widgets it will be unnecessarily confusing. No problem at all changing it.

Would another option be to convert _both_ WordPress.com and Jetpack to use the REST API?

As for the blavatar and descriptions, I'm planning on using this API to retrieve the information (the "icon" and "description" fields):
https://developer.wordpress.com/docs/api/1.2/get/sites/%24site/
This will result in potentially a lot more HTTP requests occasionally. As you're aware the data is cached via a transient, but we can discuss if we need to be more careful. I have it mostly working locally but still have some changes to make. Was that API your expectation for getting this data?

mdawaffe commented:
I'd like to avoid `IS_WPCOM` as much as possible.

Rather than have an object/hierarchy abstraction, it's typically more WordPressy to add filters/actions.

For example, you saw that `Jetpack::staticize_subdomain()` is the Jetpack equivalent to WordPress.com's `staticize_subdomain()`. Nice sleuthing :) As you can see in the Top Posts widget, Jetpack filters URLs that need staticizing:
https://github.com/Automattic/jetpack/blob/14e981306569d930253357d64f3c9826680c960a/modules/widgets/top-posts.php#L252
(Filter: https://github.com/Automattic/jetpack/blob/7b2f9f8425ddecac4e84ead5080bdc9d3afcb6e9/jetpack.php#L115)
That same code runs on WordPress.com, where, instead of attaching `Jetpack::staticize_subdomain()`, we attach `staticize_subdomain()`. The widget code is the same, but what ends up attached to the filters is different.

The stats call is also a hook, this time an action: https://github.com/Automattic/jetpack/blob/14e981306569d930253357d64f3c9826680c960a/modules/widgets/top-posts.php#L211

What do you think of this pattern?

> Would another option be to convert _both_ WordPress.com and Jetpack to use the REST API?

It's possible, but not very performant. I think it's best to keep these separate.

On WordPress.com, we also have a mechanism by which we can make REST API calls directly with PHP (that is, without making an HTTP request), but that'd still mean we'd have two different data fetching methods (one for Jetpack: HTTP, and one for WordPress.com: API via PHP).

> https://developer.wordpress.com/docs/api/1.2/get/sites/%24site/

Good idea, and good find. And, yes, that will be a lot of HTTP requests, even cached. Is there any way to get multiple sites worth of data in one request?

themoonisdown commented:
> Rather than have an object/hierarchy abstraction, it's typically more WordPressy to add filters/actions.

Thanks for this advice and the details around it. I'm totally cool using this more native pattern, and I will update the widget as such.

> Is there any way to get multiple sites worth of data in one request?

Ah yes, we can use the batch API: https://developer.wordpress.com/docs/api/1.1/get/batch/
Using this we can make one request and get all the data we need. (I also plan on limiting the response data to just the fields we are interested in.)

themoonisdown commented:
Ok, I've moved everything to actions/filters and we are now using the batch REST API to get blavatar/description/name.

@mdawaffe Could you please check over the actions/filters to make sure it is what you expected?

I have a bunch of TODOs noting where WordPress.com will have to do something hooking into the filters. (I also have a bunch of dangling TODOs for development that need to be cleaned up.) And of course we will continue the larger WP.com subscriptions discussion next week.

enejb commented on class.jetpack.php:
```diff
@@ -6354,4 +6354,13 @@ function jetpack_user_col_style() {
+	 *
+	 * @return string The blog's locale, e.g. 'en' or 'it'
+	 */
+	public static function get_blog_locale() {
```
I think this will not work as expected. The wpcom local should be calculated something like the following 
https://github.com/Automattic/jetpack/blob/master/_inc/lib/admin-pages/class.jetpack-react-page.php#L206
```
$localeSlug = explode( '_', get_locale() );
return $localeSlug[0]; 
```

Since you can't assume that the wpcom is only the first 2 characters. Since there are languages such as `Alemannisch` that use more then 2 characters to identify the language. 

But maybe I am completely wrong then you can just ignore me :)

themoonisdown replied:
I think you are probably correct. In my case this is used to get to different subdomain URLs, e.g. `en.wordpress.com`. I don't know where the master list of these URLs is kept, but seeing as further searching reveals there are many >2 character shortcodes for locales, I'm going to update the code with your suggestion.

enejb commented on modules/widgets/blogs-i-follow/blogs-i-follow.php:
```diff
@@ -0,0 +1,540 @@
+		$this->subscriptions[$this->id] = get_transient( $this->id . '-subscriptions' );
+
+		// TODO: Remove DEBUG
+		if ( WP_DEBUG || empty( $this->subscriptions[$this->id] ) ) {
```
some space between the square brackets

enejb commented on modules/widgets/blogs-i-follow/blogs-i-follow.php:
```diff
@@ -0,0 +1,540 @@
+		if ( WP_DEBUG || empty( $this->subscriptions[$this->id] ) ) {
+			delete_transient( $this->id . '-widget' );
+
+			$this->subscriptions[$this->id] = array();
```
here to..

enejb commented on modules/widgets/blogs-i-follow/blogs-i-follow.php:
```diff
@@ -0,0 +1,540 @@
+				}
+			}
+			// Update the subscriptions array with the additional fields retrieved from the API request
+			foreach ($subscriptions as &$subscription) {
```
Add some more space here.

enejb commented on modules/widgets/blogs-i-follow/blogs-i-follow.php:
```diff
@@ -0,0 +1,540 @@
+						// is not available via the REST API, so Jetpack currently cannot retrieve it.
+						$email = apply_filters( 'wpcom_blog_option', null, $subscription['blog_id'], 'admin_email' );
+						$http = is_ssl() ? 'https' : 'http';
+						$img = get_avatar( $email, self::$avatar_size, apply_filters( 'jetpack_static_url', esc_url_raw( $http . '://' . self::$default_avatar ) ) );
```
Lets document this filter so that it shows up in the developer.jetpack.com

themoonisdown replied:
If the primary intent is to facilitate code sharing between WordPress.com and Jetpack, should it still be documented in that way?

jeherve replied:
We document all filters the same way, regardless of whether the code will be ported over to WordPress.com or not. WordPress.com devs are then able to use the information provided in the Jetpack Codex, even when developing on WordPress.com.

themoonisdown replied:
@jeherve Thanks for the information! I updated the code documenting the additional filters.

enejb commented on modules/widgets/blogs-i-follow/blogs-i-follow.php:
```diff
@@ -0,0 +1,540 @@
+				$output .= "</div>";
+			}
+
+			$output .= "</div><div style='clear: both;'></div>";
```
`<div style='clear: both;'></div>` could this be done in the css?

themoonisdown replied:
Yikes...looking into this there are other problems with the CSS (particular the responsive smaller layouts) I hadn't yet seen. Thanks for spurring me into checking that out.

As for this particular line, from what I can tell so far it isn't doing anything. Other CSS already clears the floats in this grid. I think I'm going to have to make larger updates, so if it still seems unnecessary I'll delete it as part of them (or integrate it if it is needed).

themoonisdown replied:
May have spoken too soon. Testing on WP.com reveals no such bugs, so the more likely cause for what I was seeing is that my manual copy is an incomplete `widget-grid-list` stylesheet. I'll see if this clear can be removed after I get the updated CSS from @mdawaffe .

enejb commented on modules/widgets/blogs-i-follow/blogs-i-follow.php:
```diff
@@ -0,0 +1,540 @@
+		<p>
+			<label for="<?php echo $this->get_field_id( 'number' ); ?>"><?php esc_html_e( 'Number of blogs to show:', 'jetpack' ); ?>
+				<input type="number" id="<?php echo $this->get_field_id( 'number' ); ?>" name="<?php echo $this->get_field_name( 'number' ); ?>" value="<?php echo esc_attr( $instance['number'] ); ?>" min="1" max="50" />
+				<br><small><?php esc_html_e( '(at most 50)', 'jetpack' ) ?> <a href="https://en.support.wordpress.com/widgets/blogs-i-follow-widget/" target="_blank">( ? )</a></small>
```
I think there might be a better way to translate this.  (at most 50) ( ? ).
Some languages do not use spaces to differentiate between words. So things like spaces and ? should be part of the string that gets translated.

themoonisdown replied:
Do you think that '( ? )' link is useful to keep around? It doesn't provide any information or reasoning as to why the limit is 50, it just links back to the widget description page.

It is fine if we want to keep it (and we can make the necessary translation changes). But I'm unsure if it's providing any value.

themoonisdown replied:
@enejb Any thoughts on my response? Thanks!

enejb replied:
I think it would be ok to remove it. CC @lancewillett since you updated the link the .com side when it was out of date.

themoonisdown replied:
Just to clarify, are you saying we need to update the link both here and on the WP.com version of the widget? Or something else?

enejb commented on modules/widgets/blogs-i-follow/widget-grid-and-list.css:
```diff
@@ -0,0 +1,116 @@
+
+/* 2-Column Grid Layout */
+
+.widgets-grid-layout {
```
It might be good idea to prefix these classes since .widgets-grid-layout sounds pretty generic.

themoonisdown replied:
This is actually a manual copy of a CSS file that exists somewhere in the WordPress.com repo. I'm pretty sure it used to be a part of Jetpack though. I believe that I need to get the real file(s) into Jetpack and register it with the other common stylesheets. (Same thing with hover-bubbles.)

cc @mdawaffe Does that make sense, and if so, could you please provide me with those files? (Or if not, let me know how to proceed with what I've grabbed.)

themoonisdown commented on modules/widgets/blogs-i-follow/blogs-i-follow.php:
```diff
@@ -0,0 +1,540 @@
+	 * @return string the inferred blog name
+	 */
+	function get_inferred_blog_name( $subscription ) {
+		return rtrim( str_replace( 'http://', '', empty( $subscription['blog_url'] ) ? $subscription['feed_url'] : $subscription['blog_url'] ), '/' );
```
Good catch! I'll change this to consider both. The code I originally inherited used this line:

    $sub['blog_name'] = rtrim( str_replace( 'http://', '', empty( $sub['blog_url'] ) ? $sub['feed_url'] : $sub['blog_url'] ), '/' );

themoonisdown commented on modules/widgets/blogs-i-follow/blogs-i-follow.php:
```diff
@@ -0,0 +1,540 @@
+	 * a /batch REST API call
+	 */
+	private static function create_blavatar_query( $subscriptions ) {
+		$url_string = "";
```
Wow, much easier! Thanks, this dramatically simplifies that code!

themoonisdown commented:
@enejb Thanks so much for your feedback! I've committed several of your suggestions, and I've left questions on some of your other comments. Let me know what you think.

As far as testing, you're welcome to do so but be aware that things are very much still in progress 😄. We still need to determine how to retrieve the subscriptions from WordPress.com, and I'm planning on further refining and adding unit tests where appropriate.

Thanks again!

themoonisdown commented:
@jeherve Could you check out the latest changes that share the stylesheet between Top Posts and Blogs I Follow? It's related to the commit here: https://github.com/Automattic/jetpack/commit/5f8eb35a4f3024ef2123c16e172dca327d1f556a

The stylesheet is going to be used by both widgets. If there is a better place to put it (particularly the style registration), please let me know. Thanks!

jeherve commented:
@themoonisdown That should work, but I would add comments in the stylesheet, explaining where it's used and needed. I had originally moved it to a separate file because it wasn't used by any other widget, so if we change that back we should explain why (The Blogs I Follow Widget needs the styles as much as the Top Posts Widget does).

In the end it won't matter much, as all CSS is concatenated into a single `jetpack.css` file. To make sure this happens properly, you'll need to update `gulpfile.js` and add your CSS files there, as well as in the `$concatenated_style_handles` array in `class.jetpack.php`.

themoonisdown commented:
@jeherve Thanks! I added the comment as suggested. I updated the gulpfile and `$concatenated_style_handles` as well (and verified the concatenated CSS contained the styles).

themoonisdown commented:
Updated with the new direction that caches the data periodically via a cron. This will prevent against "cache miss stampedes" of HTTP requests.

Here are the unit test results. There are more to come, but I figured it would be a good idea to get feedback on the direction before I completed the thorough testing.

```
$ phpunit WP_Test_Jetpack_Blogs_I_Follow tests/php/modules/widgets/test_blogs-i-follow.php 
To run Jetpack woocommerce tests, prefix phpunit with JETPACK_TEST_WOOCOMMERCE=1
Installing...
Running as single site... To run multisite, use -c tests/phpunit/multisite.xml
Not running ajax tests. To execute these, use --group ajax.
Not running ms-files tests. To execute these, use --group ms-files.
Not running external-http tests. To execute these, use --group external-http.
PHPUnit 5.7.11 by Sebastian Bergmann and contributors.

.....                                                               5 / 5 (100%)

Time: 1.47 seconds, Memory: 26.00MB

OK (5 tests, 14 assertions)
```

themoonisdown commented:
@mdawaffe Could you take a look at the current widget, particularly the caching strategy? Thanks!

themoonisdown commented:
Just a heads up: I'm going to fix that unit test failure. It is not product code, it is a PHPUnit function unavailable in some versions of PHP apparently.

enejb commented:
Thanks for working on this @themoonisdown. I left some questions and some fixes. I still haven't had a chance to try it out fully would you say now would be a good time? :)

themoonisdown commented:
@enejb Thanks! Absolutely a good time. ;) The only thing is you will need the WP.com side of my changes as well, then add those defines I noted above. Is it possible to pull those into your own instance? Or we can coordinate and I can turn mine on and you can use the defines exactly as above.

kwight commented:
@themoonisdown hello! Are you still working on this?

themoonisdown commented:
@kwight Unfortunately, not any more. My trial ended. I am sure that others in this thread could pick it up though if desired. The necessary WP.com changes and any other relevant information is linked from my posts within the Poseidon P2.

I will close this PR for now as someone would have to pick it up and get the server side changes up and running again then create their own request. Thanks!
